lib/puppet/pops/loader/puppet_function_instantiator.rb

```diff
@@ -64,18 +65,19 @@ def self.create(loader, typed_name, source_ref, pp_code_string)
   #   typed name, and an instantiated function with global scope closure associated with the given loader
   #
   def self.create_from_model(function_definition, loader)
-    closure_scope = Puppet.lookup(:global_scope) { {} }
-    created = create_function_class(function_definition, closure_scope)
+    closure_scope = nil; # Puppet.lookup(:global_scope) { {} }
+    created = create_function_class(function_definition)
     typed_name = TypedName.new(:function, function_definition.name)
     [typed_name, created.new(closure_scope, loader)]
   end
 
-  def self.create_function_class(function_definition, closure_scope)
+  def self.create_function_class(function_definition)
     # Create a 4x function wrapper around a named closure
     Puppet::Functions.create_function(function_definition.name, Puppet::Functions::PuppetFunction) do
+      # This is highly problematic - it binds both an Evaluator and closure_scope
       init_dispatch(Evaluator::Closure::Named.new(
         function_definition.name,
-        Evaluator::EvaluatorImpl.new(), function_definition, closure_scope))
+        Evaluator::EvaluatorImpl.new(), function_definition, nil))
```

Comment:
@hlindberg, this line is the culprit that makes the test (and a puppet apply) fail. Changing this line to
```
        Evaluator::EvaluatorImpl.new(), function_definition, Puppet.lookup(:global_scope) { {} }))
```
makes it work (but does of course not fix the problem). Could it be that the global scope, at the time of the call, doesn't see the type definitions defined in the same file as the function definition? Perhaps the test is all wrong (the ducks are not in a row) since it contains a type definition, a function definition, and a call in one single file?